### DISCUSSION TITLE
Use a Github Code search as an alternate way to find a set or target repos

### DISCUSSION THREAD

**Discussion by jamestelfer** [lindell/multi-gitter#386] [Ideas]
At the moment it is possible to specify a group or a repo list to target multi-gitter at a set of repositories.

I have been using code search to discover a set of repositories to target, then feeding that list of repositories into multi-gitter. It'd be nice to cut out the middle-man and drive an action by a search directly.

Would this kind of alteration be welcome?

**Comment by lindell** [comment]
I think this would be a good thing 👍 If you want to contribute with a PR, I would be happy to merge.

**Reply by jamestelfer** [reply]
Sweet. I'll have a look and see what effort might be involved to do this. Would it be OK if it was a GitHub-only feature for a start?

**Reply by lindell** [reply]
Only GitHub is ok ☺️

**Reply by jamestelfer** [reply]
Here is a draft PR for appraisal: https://github.com/lindell/multi-gitter/pull/397

No tests as yet. I need to put some time into understanding the way tests in this repo work to get it right, but I thought a quick and dirty first look would be useful.

**Reply by jamestelfer** [reply]
This PR is a repo search by the way, rather than code search. I was thinking I'd start with what appeared to be an easier change, then go on to add `--code-search` as well.

I have very little insight into other SCM and what their needs might be, so my thoughts on `--repo-search` and `--code-search` being separate are definitely tied to the GitHub way of doing things. Hopefully it sparks discussion though.

**Reply by jamestelfer** [reply]
#398 is the issue created for the `--repo-search` enhancement.

**Reply by jamestelfer** [reply]
#399 adds `--code-search`, though this is a draft at the moment.

**Comment by josealdaco** [comment]
I think using a regular expression to search for target repositories as mentioned in #388 would be great, and can be introduced with the `--R` flag. Can this be something we can possibly contribute to?

**Reply by jamestelfer** [reply]
I've put together a PR that uses [GitHub repo search](https://github.com/lindell/multi-gitter/pull/397) if you're interested?

(I am not the maintainer, these are just my thoughts.)

I was thinking about this kind of thing myself, and headed down the search avenue first. The key drawback in my mind for the regex filter is that it occurs after the initial query, so it becomes less efficient as the size of an organization grows. I haven't done any performance analysis on this though, so I don't know how much of an issue this might be.

The downside for search is that the syntax depends on the target system: GitHub and GitLab will have different search syntaxes I'm guessing.

The big upside for search is that it can do the substring-style searches that would be common for regex, but it can also do (for GitHub at least) other things like search licenses, or README contents, or repo topics/descriptions.

The above aside, I don't see this as an either/or: it's likely that both could co-exist. If there was a `--filter-repo` flag, for example, it could filter the set of repos from whatever source: organization, repo, search, whatever.

**Reply by josealdaco** [reply]
a `--filter-repo` would be greatly beneficial to lower target repositories from using Github's search API. That said using Github's repo search query alone is quite powerful and has allowed me to target repositories by matching a repository prefix name, can't wait for your feature to officially release!
